### PR TITLE
handle inplace operations in _Feature.__torch_function__

### LINKED ISSUE #6669: inplace operations on `features._Feature`'s keeps the wrapping in tact
Going by the logic of

https://github.com/pytorch/vision/blob/7f4c55b1dac2418c594267c314afa5353b35608f/torchvision/prototype/features/_feature.py#L74-L81

that should not be happening.

```pycon
>>> image = features.Image(torch.rand(3, 16, 16))
>>> type(image.add(0))
torch.Tensor
>>> type(image.add_(0))
torchvision.prototype.features._image.Image
```

cc @bjuncek

### PR DESCRIPTION
**Pull request by pmeier**
Fixes #6669. At least as far as it is possible. We can only unwrap the result of an inplace operation, i.e. what is returned by it. We cannot change the type inplace and thus we can't prevent the following:

```pycon
>>> image = features.Image(torch.rand(3, 16, 16))
>>> output = image.add_(0)
>>> type(output)
torch.Tensor
>>> type(image)
torchvision.prototype.features._image.Image
```

As you can see `output` is properly unwrapped, but we can't do that for `image` although we have no idea if the operation "invalidated" the type.

### CODE REVIEW

File: torchvision/prototype/features/_feature.py [hunk 1]
```diff
@@ -96,18 +102,20 @@ def __torch_function__(
...
-        if not isinstance(args[0], cls):
-            return output
+            # Apart from `func` needing to be an exception, we also require the primary operand, i.e. `args[0]`, to be
+            # an instance of the class that `__torch_function__` was invoked on. The __torch_function__ protocol will
+            # invoke this method on *all* types involved in the computation by walking the MRO upwards. For example,
+            # `torch.Tensor(...).to(features.Image(...))` will invoke `features.Image.__torch_function__` with
+            # `args = (torch.Tensor(), features.Image())` first. Without this guard, the original `torch.Tensor` would
+            # be wrapped into a `features.Image`.
+            if func in cls._NO_WRAPPING_EXCEPTIONS and isinstance(args[0], cls):
+                return cls._NO_WRAPPING_EXCEPTIONS[func](cls, args[0], output)  # type: ignore[no-any-return]
```
Comment (datumbox) [inline torchvision/prototype/features/_feature.py:112]: Optional Nit: you can avoid the double lookup on the dictionary by using `.get(func, None)`.

**Review by datumbox — APPROVED** [review]
LGTM, thanks!

File: test/test_prototype_transforms_functional.py [hunk 2]
```diff
@@ -907,15 +907,15 @@ def test_correctness_gaussian_blur_image_tensor(device, image_size, dt, ksize, s
...
     inpt = torch.rand(1, 3, 32, 32)
     output = F.normalize(inpt, mean=[0.5, 0.5, 0.5], std=[1.0, 1.0, 1.0])
-    assert isinstance(output, torch.Tensor)
+    assert type(output) is torch.Tensor
     torch.testing.assert_close(inpt - 0.5, output)
 
     inpt = make_image(color_space=features.ColorSpace.RGB)
     output = F.normalize(inpt, mean=[0.5, 0.5, 0.5], std=[1.0, 1.0, 1.0])
-    assert isinstance(output, torch.Tensor)
+    assert type(output) is torch.Tensor
```
Comment (pmeier) [inline test/test_prototype_transforms_functional.py:918]: This test was wrong and this is why it didn't fail before fixing `normalize`: since `features.Image` is a subclass of `torch.Tensor` the check passes. We need to check for type identity here.

File: test/test_prototype_transforms_functional.py [hunk 3]
```diff
@@ -907,15 +907,15 @@ def test_correctness_gaussian_blur_image_tensor(device, image_size, dt, ksize, s
     torch.testing.assert_close(out, true_out, rtol=0.0, atol=1.0, msg=f"{ksize}, {sigma}")
 
 
-def test_midlevel_normalize_output_type():
+def test_normalize_output_type():
```
Comment (pmeier) [inline test/test_prototype_transforms_functional.py:910]: I renamed since

1. we agreed to use the kernel / dispatcher scheme rather than low / mid level kernels.
2. using `normalize` in the name is sufficient, since the kernel is called `normalize_image_tensor`

File: torchvision/transforms/functional_tensor.py [hunk 4]
```diff
@@ -937,8 +937,7 @@ def normalize(tensor: Tensor, mean: List[float], std: List[float], inplace: bool
         mean = mean.view(-1, 1, 1)
     if std.ndim == 1:
         std = std.view(-1, 1, 1)
-    tensor.sub_(mean).div_(std)
-    return tensor
```
Comment (pmeier) [inline torchvision/transforms/functional_tensor.py:941]: Behavior is equivalent for plain tensors, but gives us the unwrapping behavior we want from this change. With this in place, explicit unwrapping in `normalize` is no longer needed. Let me know if I should keep it there. Runtime is the same, because here we would implicitly unwrap inside `__torch_function__`.

File: torchvision/prototype/transforms/functional/_misc.py [hunk 5]
```diff
@@ -12,12 +12,17 @@
...
 ) -> torch.Tensor:
-    if not isinstance(inpt, torch.Tensor):
-        raise TypeError(f"img should be Tensor Image. Got {type(inpt)}")
+    if torch.jit.is_scripting():
+        correct_type = isinstance(inpt, torch.Tensor)
     else:
-        # Image instance after normalization is not Image anymore due to unknown data range
-        # Thus we return Tensor for input Image
-        return normalize_image_tensor(inpt, mean=mean, std=std, inplace=inplace)
+        correct_type = features.is_simple_tensor(inpt) or isinstance(inpt, features.Image)
```
Comment (pmeier) [inline torchvision/prototype/transforms/functional/_misc.py:18]: This gives is the correct error behavior in eager mode. @vfdev-5 noted that before, we could pass bounding boxes or the like to `normalize` and it would only fail in the computation rather than being caught before.

File: torchvision/transforms/functional_tensor.py (hunk 4 above)
Comment (datumbox) [reply torchvision/transforms/functional_tensor.py:941]: Initially I thought this was a styling change. Then I realize you do this to avoid returning `tensor` (which might be an image) and instead return the unwrapped `torch.Tensor`. Might worth adding a short comment in a follow up PR.
